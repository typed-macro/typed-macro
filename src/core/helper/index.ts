export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, State, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
