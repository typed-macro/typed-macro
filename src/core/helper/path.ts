import { dirname, resolve, sep } from 'path'
import { existsSync } from 'fs'

export type PathPatternNormalizer = {
  /**
   * Normalized path or path pattern,
   */
  normalized: string
  /**
   * Base path for the normalized path or path pattern.
   */
  base: string
  /**
   * Get import path from file path got by normalized path and base,
   * almost the opposite operation of normalization
   * @param path file path
   */
  resolveImportPath: (path: string) => string
}

export function normalizePathPattern(
  pattern: string,
  importer: string,
  root: string
): PathPatternNormalizer {
  if (!pattern.startsWith('.') && !pattern.startsWith('/')) {
    throw new Error(
      `pattern must start with '.' or '/' (relative to project root)`
    )
  }
  let base: string
  let parentDepth = 0
  const isAbsolute = pattern.startsWith('/')
  if (isAbsolute) {
    base = resolve(root)
    pattern = pattern.slice(1)
  } else {
    base = dirname(importer)
    while (pattern.startsWith('../')) {
      pattern = pattern.slice(3)
      base = resolve(base, '../')
      parentDepth++
    }
    if (pattern.startsWith('./')) {
      pattern = pattern.slice(2)
    }
  }
  return {
    normalized: pattern,
    base,
    resolveImportPath: isAbsolute
      ? (f) => `/${f}`
      : parentDepth
      ? (f) => `${'../'.repeat(parentDepth)}${f}`
      : (f) => `./${f}`,
  }
}

const projectDirCache = {
  root: '',
  leaf: '',
}

export function projectDir(which: 'root' | 'leaf') {
  if (projectDirCache[which]) return projectDirCache[which]

  const paths = (require.main?.path || process.cwd()).split(sep)

  if (which === 'root') {
    for (let i = 0; i < paths.length; i++) {
      const path = paths.slice(0, i + 1).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['root'] = path)
    }
  } else {
    for (let i = paths.length; i > 0; i--) {
      const path = paths.slice(0, i).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['leaf'] = path)
    }
  }

  throw new Error('can not find project root')
}
