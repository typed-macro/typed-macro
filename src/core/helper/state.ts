export type State = {
  /**
   * Get the state shared during traversal.
   */
  get<T, K>(key: K): T

  /**
   * Set the state shared during traversal.
   */
  set<T, K>(key: K, value: T): T

  /**
   * Clear the state shared during traversal.
   */
  clear(): boolean

  /**
   * Delete the state shared during traversal.
   */
  delete<K>(key: K): void
}

export function createState(): State {
  const state = new Map<any, any>()
  return {
    get(key) {
      return state.get(key)
    },
    set(key, value) {
      state.set(key, value)
      return value
    },
    clear() {
      state.clear()
      return true
    },
    delete<K>(key: K) {
      state.delete(key)
    },
  }
}
