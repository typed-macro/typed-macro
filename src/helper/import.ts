import {
  ImportDeclaration,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isImportSpecifier,
} from '@babel/types'

/**
 * ```typescript
 * import 'moduleName'
 * { moduleName: string }
 *
 * import defaultName from 'moduleName'
 * { defaultName: string; moduleName: string }
 *
 * import { exportName } from 'moduleName'
 * { exportName: string; moduleName: string }
 *
 * import { exportName as localName } from 'moduleName'
 * { localName: string; exportName: string; moduleName: string }
 *
 * import * as namespaceName from 'moduleName'
 * { namespaceName: string; moduleName: string }
 * ```
 */
export type ImportOption =
  | { moduleName: string }
  | { defaultName: string; moduleName: string }
  | { localName: string; exportName: string; moduleName: string }
  | { exportName: string; moduleName: string }
  | { namespaceName: string; moduleName: string }

export const renderImportStmt = (imp: ImportOption) =>
  'defaultName' in imp
    ? `import ${imp.defaultName} from '${imp.moduleName}'`
    : 'localName' in imp
    ? `import { ${imp.exportName} as ${imp.localName} } from '${imp.moduleName}'`
    : 'exportName' in imp
    ? `import { ${imp.exportName} } from '${imp.moduleName}'`
    : 'namespaceName' in imp
    ? `import * as ${imp.namespaceName} from '${imp.moduleName}'`
    : `import '${imp.moduleName}'`

export function matchImportStmt(
  imp: ImportOption,
  node: ImportDeclaration,
  loose = true
) {
  if (node.source.value !== imp.moduleName) return false
  if ('defaultName' in imp) {
    // import defaultName from 'moduleName'
    for (const s of node.specifiers) {
      if (isImportDefaultSpecifier(s) && s.local.name === imp.defaultName)
        return true
    }
  } else if ('localName' in imp) {
    // import { exportName as localName } from 'moduleName'
    for (const s of node.specifiers) {
      if (
        isImportSpecifier(s) &&
        s.local.name === imp.localName &&
        isIdentifier(s.imported) &&
        s.imported.name === imp.exportName
      )
        return true
    }
  } else if ('exportName' in imp) {
    // import { exportName } from 'moduleName'
    for (const s of node.specifiers) {
      if (
        isImportSpecifier(s) &&
        isIdentifier(s.imported) &&
        s.imported.name === imp.exportName &&
        (loose || s.local.name === imp.exportName)
      )
        return true
    }
  } else if ('namespaceName' in imp) {
    // import * as defaultName from 'moduleName'
    for (const s of node.specifiers) {
      if (isImportNamespaceSpecifier(s) && s.local.name === imp.namespaceName)
        return true
    }
  } else {
    // import 'moduleName'
    if (loose || node.specifiers.length === 0) return true
  }
  return false
}
