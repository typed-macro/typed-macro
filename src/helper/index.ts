export * from './macro'
export * from './babel'
export type { ImportOption } from './import'
export type { DevServerHelper } from './server'
