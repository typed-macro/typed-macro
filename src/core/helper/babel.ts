import template from '@babel/template'
import traverse from '@babel/traverse'
import * as types from '@babel/types'
import { parse, parseExpression } from '@babel/parser'
import generate from '@babel/generator'

export type Babel = {
  /**
   * From @babel/template.
   * Help generate nodes from code string.
   *
   * e.g.
   * ```typescript
   * template.statement.ast('const a = 1') // generate a statement node
   * ```
   */
  template: typeof template

  /**
   * From @babel/traverse.
   * Help traverse from a node.
   *
   * e.g.
   * ```typescript
   * traverse(someNode, {
   *   CallExpression(path) {
   *     ...
   *   }
   * })
   * ```
   */
  traverse: typeof traverse

  /**
   * From @babel/types.
   * Help check node type and create nodes of specific type.
   *
   * e.g.
   * ```typescript
   *  types.isIdentifier(someNode) // check is identifier
   *  types.identifier('hello') // create identifier node
   * ```
   */
  types: typeof types

  /**
   * From @babel/parser.
   * Help parse code string to ast.
   *
   * e.g.
   * ```typescript
   *  parse('const a = 1')
   * ```
   */
  parse: typeof parse

  /**
   * From @babel/parser.
   * Help parse code string to expression.
   *
   * e.g.
   * ```typescript
   *  parseExpression('1 + 2')
   * ```
   */
  parseExpression: typeof parseExpression

  /**
   * From @babel/generator.
   * Help generate code string from ast.
   *
   * e.g.
   * ```typescript
   *  generate(someNode)
   * ```
   */
  generate: typeof generate
}

export const BABEL_TOOLS: Babel = Object.freeze({
  template,
  traverse,
  types,
  parse,
  parseExpression,
  generate,
})
