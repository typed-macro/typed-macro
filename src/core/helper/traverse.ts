import {
  CallExpression,
  File,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isMemberExpression,
  isStringLiteral,
  Program,
} from '@babel/types'
import traverse, { NodePath, Scope } from '@babel/traverse'
import { ImportOption, matchImportStmt } from '@/core/helper/import'
import { Macro } from '@/core/macro'
import { NamespacedMacros } from '@/core/exports'
import template from '@babel/template'

export function findProgramPath(ast: File) {
  let path: NodePath<Program>
  traverse(ast, {
    Program(p) {
      path = p
      p.stop()
    },
  })
  return path!
}

export function findImported(
  program: NodePath<Program>,
  imp: ImportOption,
  loose = true
) {
  // An import declaration can only be used in top-level.
  for (const path of program.get('body') as NodePath[]) {
    if (!path.isImportDeclaration()) continue
    if (matchImportStmt(imp, path.node, loose)) return path
  }
}

export type ImportedAsNS = {
  importedAsNamespace: true
  namespace: string
  local: string
  macros: Macro[]
}

export type ImportedAsNamed = {
  importedAsNamespace: false
  namespace: string
  local: string
  // not always macro
  macro?: Macro
}

export type ImportedMacro = ImportedAsNS | ImportedAsNamed

export function findImportedMacros(
  ast: File,
  macros: NamespacedMacros,
  keepImportStmt = false
) {
  const importedMacros: ImportedMacro[] = []
  const namespaces = Object.keys(macros)

  traverse(ast, {
    ImportDeclaration(path) {
      const ns = path.node.source.value
      if (!namespaces.includes(ns)) return

      // case - import 'a'
      if (!path.node.specifiers.length) return

      path.node.specifiers.forEach((s) => {
        if (isImportDefaultSpecifier(s) || isImportNamespaceSpecifier(s)) {
          // case - import a from 'a'
          // or
          // case - import * as a from 'a'
          importedMacros.push({
            importedAsNamespace: true,
            namespace: path.node.source.value,
            local: s.local.name,
            macros: macros[ns],
          })
        } else {
          // case - import { a as a } from 'a'
          const exportName = isIdentifier(s.imported)
            ? s.imported.name
            : s.imported.value

          importedMacros.push({
            importedAsNamespace: false,
            namespace: path.node.source.value,
            local: s.local.name,
            macro: macros[ns].find((m) => m.name === exportName),
          })
        }
      })
      keepImportStmt
        ? path.replaceWith(
            template.statement.ast(`import '${path.node.source.value}'`)
          )
        : path.remove()
    },
  })

  // update scope so that macros don't have bindings now
  findProgramPath(ast).scope.crawl()

  return importedMacros
}

function isLocalDefined(scope: Scope, name: string) {
  // local-defined identifiers must have bindings,
  // macros don't have bindings,
  // but no bindings doesn't mean is a macro,
  // e.g. console.log(), `console` has no bindings too.
  return scope.getBinding(name)
}

// find macros to be applied from call expr, returns a
// - Macro: the macro to be applied
// - string: is a macro call but no macro found
// - undefined: not a macro call
type CalledMacro = Macro | string | undefined

export function getCalledMacro(
  callee: NodePath<CallExpression['callee']>,
  importedMacros: Array<ImportedMacro>
): CalledMacro {
  const node = callee.node
  if (isMemberExpression(node)) {
    const ns = node.object
    if (isIdentifier(ns)) {
      if (isLocalDefined(callee.scope, ns.name)) return
      const maybeMacro = importedMacros.find(
        (m) => m.importedAsNamespace && m.local === ns.name
      ) as ImportedAsNS
      if (!maybeMacro) return
      const method = node.property
      if (isIdentifier(method)) {
        return (
          maybeMacro.macros.find((m) => m.name === method.name) || method.name
        )
      } else if (isStringLiteral(method)) {
        // case - namespace['method']()
        return (
          maybeMacro.macros.find((m) => m.name === method.value) || method.value
        )
      }
    }
  } else if (isIdentifier(node)) {
    if (isLocalDefined(callee.scope, node.name)) return
    // case - method()
    const maybeMacro = importedMacros.find(
      (m) => !m.importedAsNamespace && m.local === node.name
    ) as ImportedAsNamed
    if (!maybeMacro) return
    return maybeMacro.macro || node.name
  }
}

export function containsMacros(
  paths: NodePath[],
  importedMacros: ImportedMacro[]
) {
  return paths.map((path) => {
    // traverse can not process the root path
    if (
      path.isCallExpression() &&
      getCalledMacro(path.get('callee'), importedMacros)
    )
      return true
    let has = false
    path.traverse({
      CallExpression(p) {
        if (getCalledMacro(p.get('callee'), importedMacros)) {
          has = true
          p.stop()
        }
      },
    })
    return has
  })
}
