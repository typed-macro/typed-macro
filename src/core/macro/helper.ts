import { ImportOption } from '@/core/helper/import'
import { Node, NodePath } from '@babel/traverse'
import { CallExpression, ImportDeclaration, Program } from '@babel/types'
import {
  normalizePathPattern,
  PathPatternNormalizer,
  projectDir,
} from '@/core/helper/path'
import {
  containsMacros,
  findImported,
  ImportedMacro,
} from '@/core/helper/traverse'
import {
  appendImports,
  appendToBody,
  prependImports,
  prependToBody,
} from '@/core/helper/transform'
import { promise } from '@/common'

export type MacroHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current file
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    pattern: string,
    root?: string,
    importer?: string
  ) => PathPatternNormalizer

  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Check whether there are unexpanded macros under the path(s).
   * @param path the node path(s) to be checked, use the arguments of the current
   * call expression by default
   */
  containsMacros: (...paths: NodePath[]) => boolean[]

  /**
   * Yield current transformer to nested macros inside current
   * call expression and then call the current macro handler again
   * if the nested macros exist.
   * In another words, expand nested macros in arguments first.
   *
   * It is recommended to call this function at the beginning of the handler
   * that needs to ensure that the arguments don't contain nested macros,
   * to reduce unnecessary repeated calculations.
   *
   * e.g.
   * ```typescript
   * ({ path }, _, { yieldToNestedMacros }) => {
   *   yieldToNestedMacros()
   *   // ...your code
   * }
   * ```
   *
   * Internally, in v0.2.x, it throws a Promise, so DO NOT call it in a try block.
   */
  yieldToNestedMacros: () => never | void
}

export function createHelper(
  thisPath: NodePath<CallExpression>,
  thisProgram: NodePath<Program>,
  thisFilepath: string,
  importedMacros: ImportedMacro[]
) {
  const helpers: MacroHelper = {
    findImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose)
    },

    hasImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose) !== undefined
    },

    prependImports: (imports, program = thisProgram) => {
      return prependImports(program, imports)
    },

    appendImports: (imports, program = thisProgram) => {
      return appendImports(program, imports)
    },

    prependToBody: (nodes, program = thisProgram) => {
      return prependToBody(program, nodes)
    },

    appendToBody: (nodes, program = thisProgram) => {
      return appendToBody(program, nodes)
    },

    normalizePathPattern: (
      pattern,
      root = projectDir('leaf'),
      importer = thisFilepath
    ) => {
      return normalizePathPattern(pattern, importer, root)
    },

    projectDir,

    containsMacros: (...paths) => {
      return containsMacros(
        paths.length ? paths : thisPath.get('arguments'),
        importedMacros
      )
    },

    yieldToNestedMacros: () => {
      if (
        containsMacros(thisPath.get('arguments'), importedMacros).includes(true)
      )
        throw promise
    },
  }

  return helpers
}
