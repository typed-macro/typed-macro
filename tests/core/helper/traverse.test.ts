import { parse } from '@babel/parser'
import {
  containsMacros,
  findImported,
  findImportedMacros,
  findProgramPath,
  getCalledMacro,
} from '@/core/helper/traverse'
import template from '@babel/template'
import { ImportOption } from '@/core/helper/import'
import {
  CallExpression,
  ExpressionStatement,
  File,
  Program,
} from '@babel/types'
import { NodePath } from '@babel/traverse'
import { getAST, getPath, matchCodeSnapshot } from '../../testutils'

let ast: File
let program: NodePath<Program>

const reset = (code = `const z = 1`) => {
  ast = getAST(code)
  program = findProgramPath(ast)
}

beforeEach(reset)

describe('findProgramPath()', () => {
  it('should work', () => {
    const ast = parse(`const a = 1`)
    const program = findProgramPath(ast)
    expect(program).not.toBeNull()
    expect(program).not.toBeUndefined()
  })

  it('should be synchronized when changed', () => {
    const ast = parse(`const a = 1`)
    const programA = findProgramPath(ast)
    const programB = findProgramPath(ast)
    programA.unshiftContainer('body', template.statement.ast('const b = 2'))
    expect(programA.get('body').length).toBe(2)
    expect(programB.get('body').length).toBe(2)
  })
})

describe('findImported()', () => {
  it('should return found node path', () => {
    reset(`
  import { a as _a } from 'a'
  import na from 'a'
  import * as _na from 'a'
  import { a } from 'a'
  import 'a'`)
    const imps: ImportOption[] = [
      { moduleName: 'a', exportName: 'a' },
      { moduleName: 'a', exportName: 'a', localName: '_a' },
      { moduleName: 'a' },
      { moduleName: 'a', defaultName: 'na' },
      { moduleName: 'a', namespaceName: '_na' },
    ]
    imps.forEach((item) =>
      expect(findImported(program, item)).not.toBeUndefined()
    )
    expect(
      findImported(program, { moduleName: 'a', exportName: 'b' })
    ).toBeUndefined()
  })

  it('should work with loose=true', () => {
    {
      reset(`import { a } from 'a'`)
      expect(
        findImported(program, { moduleName: 'a', exportName: 'a' }, true)
      ).not.toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a' }, true)
      ).not.toBeUndefined()
    }
    {
      reset(`import { a as _a } from 'a'`)
      expect(
        findImported(
          program,
          {
            moduleName: 'a',
            exportName: 'a',
            localName: '_a',
          },
          true
        )
      ).not.toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a', exportName: 'a' }, true)
      ).not.toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a' }, true)
      ).not.toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a', exportName: 'b' }, true)
      ).toBeUndefined()
    }
  })

  it('should work with loose=false', () => {
    {
      reset(`import { a } from 'a'`)
      expect(
        findImported(program, { moduleName: 'a', exportName: 'a' }, false)
      ).not.toBeUndefined()
      expect(findImported(program, { moduleName: 'a' }, false)).toBeUndefined()
    }
    {
      reset(`import { a as _a } from 'a'`)
      expect(
        findImported(
          program,
          {
            moduleName: 'a',
            exportName: 'a',
            localName: '_a',
          },
          false
        )
      ).not.toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a', exportName: 'a' }, false)
      ).toBeUndefined()
      expect(findImported(program, { moduleName: 'a' }, false)).toBeUndefined()
      expect(
        findImported(program, { moduleName: 'a', exportName: 'b' }, false)
      ).toBeUndefined()
    }
  })
})

describe('findImportedMacros()', () => {
  it('should work', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import { b } from '@b'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      expect(
        findImportedMacros(getAST(c.code), c.macros as any)
      ).toMatchSnapshot()
    })
  })

  it('should rewrite import stmts if keep=true', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      findImportedMacros(ast, c.macros as any, true)
      matchCodeSnapshot(ast)
    })
  })

  it('should remove import stmts if keep=false', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      findImportedMacros(ast, c.macros as any)
      expect(ast.program.body.length).toBe(0)
    })
  })
})

describe('getCalledMacro()', () => {
  it('should work', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
      call: string
    }[] = [
      {
        code: `import { a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `_a()`,
      },
      {
        code: `import a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a.a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a['a']()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'b' }] },
        call: `a.a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'b' }] },
        call: `a['a']()`,
      },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      const importedMacros = findImportedMacros(ast, c.macros as any)

      expect(
        getCalledMacro(
          (
            (
              getPath(c.call).get('body')[0] as NodePath<ExpressionStatement>
            ).get('expression') as NodePath<CallExpression>
          ).get('callee'),
          importedMacros
        )
      ).toMatchSnapshot()
    })
  })
})

describe('containsMacros()', () => {
  it('should work', () => {
    const ast = getAST(`
  import { a } from '@a'
  a(a(), c(), c(a()))`)
    const importedMacros = findImportedMacros(
      ast,
      {
        '@a': [{ name: 'a' } as any],
      },
      true
    )
    const program = findProgramPath(ast)
    const paths = (
      (program.get('body')[1] as NodePath<ExpressionStatement>).get(
        'expression'
      ) as NodePath<CallExpression>
    ).get('arguments')
    expect(containsMacros(paths, importedMacros)).toEqual([true, false, true])
  })
})
