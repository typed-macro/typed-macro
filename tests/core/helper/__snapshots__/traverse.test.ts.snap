// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findImportedMacros() should rewrite import stmts if keep=true 1`] = `"import '@a';"`;

exports[`findImportedMacros() should rewrite import stmts if keep=true 2`] = `"import '@a';"`;

exports[`findImportedMacros() should rewrite import stmts if keep=true 3`] = `"import '@a';"`;

exports[`findImportedMacros() should rewrite import stmts if keep=true 4`] = `"import '@a';"`;

exports[`findImportedMacros() should work 1`] = `
Array [
  Object {
    "importedAsNamespace": false,
    "local": "a",
    "macro": Object {
      "name": "a",
    },
    "namespace": "@a",
  },
]
`;

exports[`findImportedMacros() should work 2`] = `
Array [
  Object {
    "importedAsNamespace": false,
    "local": "_a",
    "macro": Object {
      "name": "a",
    },
    "namespace": "@a",
  },
]
`;

exports[`findImportedMacros() should work 3`] = `
Array [
  Object {
    "importedAsNamespace": true,
    "local": "a",
    "macros": Array [
      Object {
        "name": "a",
      },
    ],
    "namespace": "@a",
  },
]
`;

exports[`findImportedMacros() should work 4`] = `
Array [
  Object {
    "importedAsNamespace": true,
    "local": "a",
    "macros": Array [
      Object {
        "name": "a",
      },
    ],
    "namespace": "@a",
  },
]
`;

exports[`findImportedMacros() should work 5`] = `Array []`;

exports[`getCalledMacro() should work 1`] = `
Object {
  "name": "a",
}
`;

exports[`getCalledMacro() should work 2`] = `undefined`;

exports[`getCalledMacro() should work 3`] = `
Object {
  "name": "a",
}
`;

exports[`getCalledMacro() should work 4`] = `undefined`;

exports[`getCalledMacro() should work 5`] = `undefined`;

exports[`getCalledMacro() should work 6`] = `
Object {
  "name": "a",
}
`;

exports[`getCalledMacro() should work 7`] = `
Object {
  "name": "a",
}
`;

exports[`getCalledMacro() should work 8`] = `"a"`;

exports[`getCalledMacro() should work 9`] = `"a"`;
